scenario/ops_main_mock.py

```diff
@@ -77,6 +80,25 @@ def main(
         if pre_event:
             pre_event(charm)
 
+        if not getattr(charm.on, dispatcher.event_name, None):
```

Comment:
Would this take custom relation events into account?

Reply:
Yes and no:

self.on.custom --> YES
self.charm_lib.on.custom --> NO

reason: ops.main fires an event on the charm if it can find it in the charm's `on`. So ops.main would not be able to fire a custom event owned by other Objects, only custom events owned by the CharmBase type being tested.

That is something we could work around, but this would be a different way of executing a charm, one that's in practice not supported by `ops`. Maybe it adds value though, it is something we can consider.

Reply:
added a test to verify

Reply:
So with this change we'd get warnings for practically any charm?
I.e. if we try to trigger in traefik events such as ipa_events.data_provided.

Reply:
This change seems good to me (it seems like a good idea to raise an exception in this case), but the `on_no_event_handler` bit seems overcomplicated to me. Why not just always raise? If the test really wants to do something different, they can catch the exception (which maybe could be a new `NoObserversError` exception subclass).

Why would you trigger an event in your test if the charm doesn't support it?

Reply:
@benhoyt good point; my reasoning was: suppose that some charm does something weird in `__init__`, which at the moment `ops` has no way to prevent, such as setting status=active.
The charm will be initialized, even though no event will be fired on it, and we'd get an error.
However we'd get the error after the init code has run, so we could catch it and do our assertions. So yeah, I'll get rid of that.

@sed-i  yes, at the moment you can't trigger `ipa_events.data_provided`, because that event doesn't belong to the charm, but to `charm.ipa`. We'd need some way to specify who the owner of the event we want to fire is, and I reckon that's a separate use case.
We could for example do:
```python
State().trigger(event=Event('data_provided', owner_path=['ipa']), charm_cls=MyCharm)
```
and then our `main` would:
```
getattr(getattr(charm_cls(), 'ipa').on, 'data_provided').emit()
```

and things from there would proceed 'as normal'. But this risks diverging with a 'true-to-juju` charm execution, so I think it has low priority.